### PR TITLE
Implement SDL 0029 Multiple File Manager Transactions

### LINKED ISSUE #558: [SDL 0029] Multiple File Manager Transactions
Proposal: [Multiple File Manager Transactions](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0029-ios-multiple-file-manager-transactions.md)

> This proposal seeks to add APIs that allow developers to make multiple file manager transactions simultaneously. For example, a developer should be able to delete or upload multiple files simultaneously and have the completion block only be fired when all have completed.

Review: smartdevicelink/sdl_evolution#91

SDLC Steering Committee Decision:

> There was full agreement from the Steering Committee to accept this proposal now that it’s been revised per the discussion between Joel Fischer (Livio) and Kutjim Shala (Ford).

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #558 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit test cases were added for each new method.

### Summary
Implements [SDL 0029](https://github.com/smartdevicelink/sdl_ios/issues/558)

This proposal adds APIs that allow developers to make multiple file manager transactions simultaneously. A developer can now delete or upload multiple files simultaneously and have the completion block only be fired when all have completed.

##### Enhancements
* Added `uploadFiles:completionHandler:`, `uploadFiles:progressHandler:completionHandler:`, and `deleteRemoteFilesWithNames:completionHandler:`  methods to the `SDLFileManager` class.
* Added test cases to `SDLFileManagerSpec` for testing the above methods. Added the classes `TestMultipleFilesConnectionManager`, `TestProgressResponse`,  and `TestResponse` to help with testing. Test cases were also added to `SDLFileSpec` to test the `inputStream` and `fileSize` parameters. 
* Added custom errors and exceptions to the `SDLError` class.
* Added examples of how to use each new method to the **SmartDeviceLink-Example** app.

##### Bug Fixes
* Fixed the `sdl_sendPutFiles:mtuSize:completion:` method in the`SDLUploadFileOperation` class so the input stream is now closed properly when it is no longer needed. This prevents memory errors when multiple files are being uploaded.
* Fixed the `deleteRemoteFileWithName:completionHandler` method in the `SDLFileManager` class so it now returns the correct error and success result. Previously it was returning `success = true` and `error = nil` even when the delete was **not** successful.
* Removed the `sdl_deleteTemporaryFile:fileURL` function in the `SDLFileManager` class. When creating a `SDLFile` using a `fileURL`, the data in the file was being deleted after the file was uploaded. Since the file data was deleted, it prevented a file being uploaded more than once during an app session (this could be problematic if the user deletes a file on the head unit and then want to reload the same file during the same app session). 
* Fixed a bug in the `deleteRemoteFileWithName:name :handler` in the `SDLFileManager` class. A `SDLDeleteFileOperation` RPC was being sent to Core even though it already determined that the image was not on Core.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLFileManager.m [hunk 1]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
+                                                                                     handler(success, self.bytesAvailable, error);
                                                                                  }
                                                                              }];
 
     [self.transactionQueue addOperation:deleteOperation];
 }
 
+- (void)deleteRemoteFilesWithNames:(NSArray<SDLFileName *> *)names completionHandler:(nullable SDLFileManagerMultiDeleteCompletionHandler)completionHandler {
+    if (names.count == 0) {
+        return completionHandler([NSError sdl_fileManager_noFilesError]);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLFileManager.m]: We should throw an exception here instead. This is a programmer error, not one that occurs "accidentally" in production.

File: SmartDeviceLink/SDLFileManager.m [hunk 2]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
 
 #pragma mark - Uploading
 
+- (void)uploadFiles:(NSArray<SDLFile *> *)files completionHandler:(nullable SDLFileManagerMultiUploadCompletionHandler)completionHandler {
+    [self uploadFiles:files progressHandler:nil completionHandler:completionHandler];
+}
+
+- (void)uploadFiles:(NSArray<SDLFile *> *)files progressHandler:(nullable SDLFileManagerMultiUploadProgressHandler)progressHandler completionHandler:(nullable SDLFileManagerMultiUploadCompletionHandler)completionHandler {
+    if (files.count == 0) {
+        return completionHandler([NSError sdl_fileManager_noFilesError]);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLFileManager.m]: We should throw an exception here instead. This is a programmer error, not one that occurs "accidentally" in production.

File: SmartDeviceLink/SDLFileManager.m (hunk 1 above)
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLFileManager.m]: Fixed. I replaced the error with an exception.

File: SmartDeviceLink/SDLFileManager.m (hunk 2 above)
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLFileManager.m]: Fixed. I replaced the error with an exception.

**Comment by NicoleYarroch** [comment]
I replaced all the test assets with SDL owned images. The soft button icons used in the SDL-Example app have been replaced with simple shape images created in Illustrator by moi.

File: SmartDeviceLinkTests/SDLFileManagerMultipleFilesSpec.m [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+#import "SDLNotificationConstants.h"
+#import "SDLPutFile.h"
+#import "SDLPutFileResponse.h"
+#import "SDLRPCResponse.h"
+#import "SDLUploadFileOperation.h"
+#import "TestMultipleFilesConnectionManager.h"
+#import "TestProgressResponse.h"
+#import "TestResponse.h"
+
+QuickSpecBegin(SDLFileManagerMultipleFilesSpec)
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/SDLFileManagerMultipleFilesSpec.m]: I know it's ugly, but we only do one spec per class, so this should go into the `SDLFileManagerSpec` as a new describe at the bottom.

File: SmartDeviceLink/SDLError.h [hunk 4]
```diff
@@ -39,7 +39,11 @@ extern SDLErrorDomain *const SDLErrorDomainFileManager;
 + (NSError *)sdl_fileManager_noKnownFileError;
 + (NSError *)sdl_fileManager_unableToStartError;
 + (NSError *)sdl_fileManager_unableToUploadError;
++ (NSError *)sdl_fileManager_unableToUploadError:(NSDictionary *)userInfo;
++ (NSError *)sdl_fileManager_unableToDeleteError:(NSDictionary *)userInfo;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLError.h]: Fixed.

File: SmartDeviceLink/SDLErrorConstants.h [hunk 5]
```diff
@@ -55,15 +55,27 @@ typedef NS_ENUM(NSInteger, SDLFileManagerError) {
...
-     * The file manager could not find the local file
+     *  The file manager could not find the local file.
      */
     SDLFileManagerErrorFileDoesNotExist = -5,
+    /**
+     *  The file manager was not passed a file.
+     */
+    SDLFileManagerErrorNoFiles = -6,
+    /*
+     *  The file upload was cancelled.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLErrorConstants.h]: Fixed

File: SmartDeviceLink/SDLFileManager.h [hunk 6]
```diff
@@ -96,6 +104,23 @@ typedef void (^SDLFileManagerStartupCompletionHandler)(BOOL success, NSError *__
 - (void)uploadFile:(SDLFile *)file completionHandler:(nullable SDLFileManagerUploadCompletionHandler)completion;
 
 /**
+ *  Uploads an array of files to the remote file system.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.h]: Fixed

File: SmartDeviceLink/SDLFileManager.h [hunk 7]
```diff
@@ -96,6 +104,23 @@ typedef void (^SDLFileManagerStartupCompletionHandler)(BOOL success, NSError *__
...
+ *
+ *  @param files An array of SDLFiles to be sent
+ *  @param progressHandler An optional completion handler that sends a response for each uploaded file. The progress handler will contain the name of the last uploaded file, the percentage of total file data uploaded, whether or not to cancel all future uploads, and the error if the file failed to upload.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.h]: Fixed

File: SmartDeviceLink/SDLFileManager.m [hunk 8]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
+    [self uploadFiles:files progressHandler:nil completionHandler:completionHandler];
+}
+
+- (void)uploadFiles:(NSArray<SDLFile *> *)files progressHandler:(nullable SDLFileManagerMultiUploadProgressHandler)progressHandler completionHandler:(nullable SDLFileManagerMultiUploadCompletionHandler)completionHandler {
+    if (files.count == 0) {
+        @throw [NSException sdl_missingFilesException];
+    }
+
+    NSMutableDictionary *failedUploads = [[NSMutableDictionary alloc] init];
+    float totalBytesToUpload = progressHandler == nil ? 0.0 : [self sdl_totalBytesToUpload:files];
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.m]: Fixed

File: SmartDeviceLink/SDLFileManager.m [hunk 9]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
+    __block float totalBytesUploaded = 0.0;
+    
+
+    dispatch_group_t uploadFilesTask = dispatch_group_create();
+    dispatch_group_enter(uploadFilesTask);
+    for(SDLFile *file in files) {
+        dispatch_group_enter(uploadFilesTask);
+
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.m]: Fixed

File: SmartDeviceLink/SDLFileManager.m [hunk 10]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
+        [self uploadFile:file completionHandler:^(BOOL success, NSUInteger bytesAvailable, NSError * _Nullable error) {
+            if(!success) {
+                failedUploads[file.name] = error;
+            }
+
+            // Send an update for each file sent to the remote
+            if (progressHandler != nil) {
+                totalBytesUploaded += file.fileSize;
+                float uploadPercentage = [self sdl_uploadPercentage:totalBytesToUpload uploadedBytes:totalBytesUploaded];
+                Boolean cancel = false;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.m]: Fixed

File: SmartDeviceLink/SDLFileManagerConstants.h [hunk 11]
```diff
@@ -10,8 +10,17 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+typedef NSString SDLFileName;
+
 typedef void (^SDLFileManagerUploadCompletionHandler)(BOOL success, NSUInteger bytesAvailable, NSError *__nullable error);
+typedef void(^SDLFileManagerMultiUploadCompletionHandler)(NSError *__nullable error);
+typedef void(^SDLFileManagerMultiUploadProgressHandler)(SDLFileName *fileName, float uploadPercentage, BOOL *cancel, NSError *__nullable error);
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLFileManagerConstants.h]: The progress handler should be altered to match the [RPC sequential send proposal design](https://github.com/joeljfischer/sdl_evolution/blob/2b41868dc7a666ab498dd5643a2e068bb5b96134/proposals/nnnn-send-multiple-rpcs-ios.md#proposed-solution) due to the current design's difficulty of compatibility with Swift.

File: SmartDeviceLinkTests/SDLFileManagerMultipleFilesSpec.m (hunk 3 above)
Comment (NicoleYarroch) [reply SmartDeviceLinkTests/SDLFileManagerMultipleFilesSpec.m]: Fixed.

File: SmartDeviceLink/SDLFileManagerConstants.h (hunk 11 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManagerConstants.h]: Changed. Removed `cancel` as a parameter. The remaining uploads are now only canceled when NO is returned in the progress handler.

File: SmartDeviceLink/SDLFileManager.m [hunk 12]
```diff
@@ -209,16 +210,126 @@ - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable
...
+                progressHandler(file.name, uploadPercentage, &cancel, error);
+                if (cancel) {
+                    // If user sets cancel to YES, cancel all future operations
+                    dispatch_group_leave(uploadFilesTask);
+                    [self.transactionQueue cancelAllOperations];
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.m]: Fixed. Now only files in same upload group are canceled.

File: SmartDeviceLink/SDLFileManagerConstants.h [hunk 13]
```diff
@@ -11,7 +11,14 @@
 NS_ASSUME_NONNULL_BEGIN
 
 typedef void (^SDLFileManagerUploadCompletionHandler)(BOOL success, NSUInteger bytesAvailable, NSError *__nullable error);
+typedef void(^SDLFileManagerMultiUploadCompletionHandler)(NSError *__nullable error);
+typedef void(^SDLFileManagerMultiUploadProgressHandler)(NSString *fileName, float uploadPercentage, Boolean *cancel, NSError *__nullable error);
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManagerConstants.h]: Changed.

File: SmartDeviceLink/SDLError.h [hunk 14]
```diff
@@ -39,14 +39,19 @@ extern SDLErrorDomain *const SDLErrorDomainFileManager;
 + (NSError *)sdl_fileManager_noKnownFileError;
 + (NSError *)sdl_fileManager_unableToStartError;
 + (NSError *)sdl_fileManager_unableToUploadError;
++ (NSError *)sdl_fileManager_unableTo__ErrorWithUserInfo:(NSDictionary *)userInfo;
+// + (NSError *)sdl_fileManager_unableToUploadError:(NSDictionary *)userInfo;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLError.h]: Fixed removed

File: SmartDeviceLink/SDLFileManager.h [hunk 15]
```diff
@@ -88,6 +88,14 @@ typedef void (^SDLFileManagerStartupCompletionHandler)(BOOL success, NSError *__
 - (void)deleteRemoteFileWithName:(SDLFileName *)name completionHandler:(nullable SDLFileManagerDeleteCompletionHandler)completion;
 
 /**
+ *  Deletes an array of files on the remote file system. The files are deleted in the order in which they are added to the array, with the first file to be deleted at index 0. The delete queue is serial, meaning that once a delete request is sent to Core, the queue waits until a response is received from Core before the next the next delete request is sent.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.h]: Fixed

File: SmartDeviceLink/SDLFileManager.h [hunk 16]
```diff
@@ -96,6 +104,25 @@ typedef void (^SDLFileManagerStartupCompletionHandler)(BOOL success, NSError *__
 - (void)uploadFile:(SDLFile *)file completionHandler:(nullable SDLFileManagerUploadCompletionHandler)completion;
 
 /**
+ *  Uploads an array of files to the remote file system. The files will be uploaded in the order in which they are added to the array, with the first file to be uploaded at index 0. The upload queue is serial, meaning that once a upload request is sent to Core, the queue waits until a response is received from Core before the next the next upload request is sent.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.h]: Fixed

File: SmartDeviceLink/SDLFileManager.h [hunk 17]
```diff
@@ -96,6 +104,25 @@ typedef void (^SDLFileManagerStartupCompletionHandler)(BOOL success, NSError *__
...
+ *
+ *  The optional progress handler can be used to keep track of the upload progress. After each file upload, the progress handler returns the upload percentage and an error, if one occured during the upload process. The progress handler also includes an option to cancel the upload of all remaining files in queue.
+ *
+ *  @param files  An array of SDLFiles to be sent
+ *  @param progressHandler  An optional block that sends a response for each uploaded file. The fileName parameter contains the name of the last uploaded file. The uploadPercentage parameter contains the percentage, as a decimal value between 0.0 and 1.0, of the total file data uploaded. The upload percentage is calculated as the total file size of all attempted file uploads (irregardless of the successfulness of the upload) divided by the sum of the data in all the files. The error parameter contains a custom error if the file failed to upload or nil if the upload was successful. The cancel parameter should be set to true if all remaining file uploads in the queue need to be canceled. If cancel is left as false, all uploads will continue as normal.
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLFileManager.h]: Fixed

File: SmartDeviceLink/SDLError.h [hunk 18]
```diff
@@ -39,14 +39,19 @@ extern SDLErrorDomain *const SDLErrorDomainFileManager;
 + (NSError *)sdl_fileManager_noKnownFileError;
 + (NSError *)sdl_fileManager_unableToStartError;
 + (NSError *)sdl_fileManager_unableToUploadError;
++ (NSError *)sdl_fileManager_unableTo__ErrorWithUserInfo:(NSDictionary *)userInfo;
```
[the remark this answers is not in the document]
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLError.h]: Fixed

**Pull request merged**